bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+      }
+    else if (input_ringbuffer_.get_writable_frames() >= n_frames && output_ringbuffer_.get_readable_frames() >= n_frames)
+      {
+        /* handle input ports */
+        uint frames_written = input_ringbuffer_.write (n_frames, in_values);
+        assert_return (frames_written == n_frames, 0); // we checked the available space before
+
+        /* handle output ports */
+        uint read_frames = output_ringbuffer_.read (n_frames, out_values);
+        assert_return (read_frames == n_frames, 0); // we checked the available space before
+      }
+    else
+      {
+        /* underrun (less than n_frames available in input/output ringbuffer) -> write zeros */
+        atomic_xruns_++;
+
+        for (auto values : out_values)
+          Block::fill (n_frames, values, 0.0);
+      }
+    return 0;
+  }
+
+  static jack_latency_range_t
+  get_latency_for_ports (const vector<jack_port_t *>& ports,
+                         jack_latency_callback_mode_t mode)
+  {
+    jack_latency_range_t range;
+
+    // compute minimum possible and maximum possible latency over all ports
+    for (size_t p = 0; p < ports.size(); p++)
+      {
+        jack_latency_range_t port_range;
+
+        jack_port_get_latency_range (ports[p], mode, &port_range);
+
+        if (!p) // first port
+          range = port_range;
+        else
+          {
+            range.min = std::min (range.min, port_range.min);
+            range.max = std::max (range.max, port_range.max);
+          }
+      }
+    return range;
+  }
+  static void
+  static_latency_callback (jack_latency_callback_mode_t  mode,
+                           void                         *jack_handle)
+  {
+    static_cast <JackPcmDriver *> (jack_handle)->latency_callback (mode);
+  }
+  void
+  latency_callback (jack_latency_callback_mode_t mode)
+  {
+    // the capture/playback latency added is the number of samples in the ringbuffer
+    if (mode == JackCaptureLatency)
+      {
+        jack_latency_range_t range = get_latency_for_ports (input_ports_, mode);
+        range.min += buffer_frames_;
+        range.max += buffer_frames_;
+
+        for (auto port : output_ports_)
+          jack_port_set_latency_range (port, mode, &range);
+      }
+    else
+      {
+        jack_latency_range_t range = get_latency_for_ports (output_ports_, mode);
+        range.min += buffer_frames_;
+        range.max += buffer_frames_;
+
+        for (auto port : input_ports_)
+          jack_port_set_latency_range (port, mode, &range);
+      }
+  }
+  static void
+  static_shutdown_callback (void *jack_handle)
+  {
+    static_cast<JackPcmDriver *> (jack_handle)->shutdown_callback();
+  }
+  void
+  shutdown_callback()
+  {
+    is_down_ = true;
+  }
+public:
+  explicit      JackPcmDriver (const String &devid) : PcmDriver (devid) {}
+  static PcmDriverP
+  create (const String &devid)
+  {
+    auto pdriverp = std::make_shared<JackPcmDriver> (devid);
+    return pdriverp;
+  }
+  ~JackPcmDriver()
+  {
+  }
```

Comment:
The dtor should release all resources, e.g. call close(), either unconditionally, or with `if (jack_client_)`. It may also be executed without open() being called previously, or after a failing open() without close().

Reply:
Right. Fixed.